Firestore/core/test/firebase/firestore/local/mutation_queue_test.cc

```diff
@@ -0,0 +1,432 @@
+    for (const MutationBatch& batch : batches) {
+      absl::optional<MutationBatch> found =
+          mutation_queue_->LookupMutationBatch(batch.batch_id());
+      ASSERT_EQ(found->batch_id(), batch.batch_id());
+    }
+
+    // Even on a nonempty queue searching should not find a non-existent batch
+    not_found = mutation_queue_->LookupMutationBatch(42);
+    ASSERT_EQ(not_found, absl::nullopt);
+  });
+}
+
+TEST_P(MutationQueueTest, NextMutationBatchAfterBatchID) {
+  persistence_->Run("test_next_mutation_batch_after_batch_id", [&] {
+    std::vector<MutationBatch> batches = CreateBatches(10);
+    std::vector<MutationBatch> removed = RemoveFirstBatches(3, &batches);
+
+    for (size_t i = 0; i < batches.size() - 1; i++) {
+      const MutationBatch& current = batches[i];
+      const MutationBatch& next = batches[i + 1];
+      absl::optional<MutationBatch> found =
+          mutation_queue_->NextMutationBatchAfterBatchId(current.batch_id());
+      ASSERT_EQ(found->batch_id(), next.batch_id());
+    }
+
+    for (size_t i = 0; i < removed.size(); i++) {
+      const MutationBatch& current = removed[i];
+      const MutationBatch& next = batches[0];
+      absl::optional<MutationBatch> found =
+          mutation_queue_->NextMutationBatchAfterBatchId(current.batch_id());
+      ASSERT_EQ(found->batch_id(), next.batch_id());
+    }
+
+    const MutationBatch& first = batches[0];
+    absl::optional<MutationBatch> found =
+        mutation_queue_->NextMutationBatchAfterBatchId(first.batch_id() - 42);
+    ASSERT_EQ(found->batch_id(), first.batch_id());
+
+    const MutationBatch& last = batches[batches.size() - 1];
+    absl::optional<MutationBatch> not_found =
+        mutation_queue_->NextMutationBatchAfterBatchId(last.batch_id());
+    ASSERT_EQ(not_found, absl::nullopt);
+  });
+}
+
+TEST_P(MutationQueueTest, AllMutationBatchesAffectingDocumentKey) {
+  persistence_->Run("test_all_mutation_batches_affecting_document_key", [&] {
+    std::vector<Mutation> mutations = {
+        testutil::SetMutation("foi/bar", Map("a", 1)),
+        testutil::SetMutation("foo/bar", Map("a", 1)),
+        testutil::PatchMutation("foo/bar", Map("b", 1), {}),
+        testutil::SetMutation("foo/bar/suffix/key", Map("a", 1)),
+        testutil::SetMutation("foo/baz", Map("a", 1)),
+        testutil::SetMutation("food/bar", Map("a", 1)),
+    };
+
+    // Store all the mutations.
+    std::vector<MutationBatch> batches;
+    for (const Mutation& mutation : mutations) {
+      MutationBatch batch =
+          mutation_queue_->AddMutationBatch(Timestamp::Now(), {}, {mutation});
+      batches.push_back(batch);
+    }
+
+    std::vector<MutationBatch> expected{batches[1], batches[2]};
+    std::vector<MutationBatch> matches =
+        mutation_queue_->AllMutationBatchesAffectingDocumentKey(
+            testutil::Key("foo/bar"));
+
+    ASSERT_EQ(matches, expected);
+  });
+}
+
+TEST_P(MutationQueueTest, AllMutationBatchesAffectingDocumentKeys) {
```

Comment:
Can this test's name be made less similar to the preceding test? It took me three readings of each name before I could spot the difference.

Reply:
Done: "AffectingMultipleDocumentKeys".